src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -412,21 +415,20 @@ public static unsafe void AddScalarU(float scalar, Span<float> dst)
         {
             fixed (float* pdst = dst)
             {
-                float* pDstEnd = pdst + dst.Length;
-                float* pDstCurrent = pdst;
-
                 Vector128<float> scalarVector = Sse.SetAllVector128(scalar);
+                int count = Math.DivRem(dst.Length, SseAlignment, out int remainder);
+                float* pDstCurrent = pdst;
 
-                while (pDstCurrent + 4 <= pDstEnd)
+                for (int i = 0; i < count; i++)
                 {
                     Vector128<float> dstVector = Sse.LoadVector128(pDstCurrent);
                     dstVector = Sse.Add(dstVector, scalarVector);
                     Sse.Store(pDstCurrent, dstVector);
 
-                    pDstCurrent += 4;
+                    pDstCurrent += SseAlignment;
                 }
 
-                while (pDstCurrent < pDstEnd)
+                for (int i = 0; i < remainder; i++)
                 {
                     Vector128<float> dstVector = Sse.LoadScalarVector128(pDstCurrent);
```

Comment:
This would be way more readable as `pDstCurrent[i] += scalar`, and it should produce the same code.

The various `scalar` intrinsics are really meant for places where you have to interop between Vector and Scalar code, or where you need some scalar operations which aren't expressible in normal C# code.

Reply:
Yes, perf test results show that this change improves the runtime significantly:
## After all changes, including changing scalar intrinsics to indexed code:
```ini
BenchmarkDotNet=v0.11.1, OS=Windows 10.0.17134.228 (1803/April2018Update/Redstone4)
Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-alpha1-20180720-2
  [Host] : .NET Core 3.0.0-preview1-26710-03 (CoreCLR 4.6.26710.05, CoreFX 4.6.26708.04), 64bit RyuJIT

Toolchain=InProcessToolchain
```
|                   Type |     Method |     Mean |    Error |   StdDev |
|----------------------- |----------- |---------:|---------:|---------:|
|    AvxPerformanceTests | AddScalarU | 172.1 us | 2.589 us | 2.422 us |
| NativePerformanceTests | AddScalarU | 216.9 us | 1.785 us | 1.491 us |
|    SsePerformanceTests | AddScalarU | 209.7 us | 1.492 us | 1.246 us |

## After partial changes (removing the 2nd time of Math.DivRem):
``` ini
BenchmarkDotNet=v0.11.1, OS=Windows 10.0.17134.228 (1803/April2018Update/Redstone4)
Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-alpha1-20180720-2
  [Host] : .NET Core 3.0.0-preview1-26710-03 (CoreCLR 4.6.26710.05, CoreFX 4.6.26708.04), 64bit RyuJIT

Toolchain=InProcessToolchain
```
|                   Type |     Method |     Mean |     Error |    StdDev |
|----------------------- |----------- |---------:|----------:|----------:|
|    AvxPerformanceTests | AddScalarU | 193.3 us | 3.6653 us | 3.4285 us |
| NativePerformanceTests | AddScalarU | 215.4 us | 0.7014 us | 0.6218 us |
|    SsePerformanceTests | AddScalarU | 238.2 us | 4.1955 us | 3.5034 us |

## Before the change:
``` ini
BenchmarkDotNet=v0.11.1, OS=Windows 10.0.17134.228 (1803/April2018Update/Redstone4)
Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-alpha1-20180720-2
  [Host] : .NET Core 3.0.0-preview1-26710-03 (CoreCLR 4.6.26710.05, CoreFX 4.6.26708.04), 64bit RyuJIT

Toolchain=InProcessToolchain
```
|                   Type |     Method |     Mean |    Error |   StdDev |
|----------------------- |----------- |---------:|---------:|---------:|
|    AvxPerformanceTests | AddScalarU | 200.3 us | 3.919 us | 4.513 us |
| NativePerformanceTests | AddScalarU | 249.4 us | 4.400 us | 4.116 us |
|    SsePerformanceTests | AddScalarU | 235.3 us | 1.393 us | 1.235 us |